### PR TITLE
MAT-6709: add endpoint and translation to get QRDA xmls

### PR DESCRIPTION
**Pull request by chubert-sb**
## MADiE PR

Jira Ticket: [MAT-6709](https://jira.cms.gov/browse/MAT-6709)
(Optional) Related Tickets:

### All Submissions
* [ ] This PR has the JIRA linked.
* [ ] Required tests are included.
* [ ] No extemporaneous files are included (i.e Complied files or testing results).
* [ ] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/gov/cms/madie/resources/PackageController.java [hunk 1]
```diff
@@ -31,6 +27,7 @@ public class PackageController {
   private final HqmfService hqmfService;
   private final TranslationServiceClient translationServiceClient;
   private final HumanReadableService humanReadableService;
+  private final QrdaService qrdaService;
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/resources/PackageController.java]: Does this get used in this class?

Comment (chubert-sb) [reply src/main/java/gov/cms/madie/resources/PackageController.java]: Missed this during cleanup. fixed

File: src/main/java/gov/cms/madie/services/QrdaService.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    return client.getQRDA(dto, accessToken);
+  }
+
+  private Map<String, Object> buildOptions(QdmMeasure measure) {
+    Long startTime =
+        Long.parseLong(
+            measure
+                .getMeasurementPeriodStart()
+                .toInstant()
+                .atZone(ZoneId.systemDefault())
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/services/QrdaService.java]: Need to remember to verify the timezone used in prod.

**Review by jkotanchik-SB — DISMISSED** [review]

File: src/main/java/gov/cms/madie/qrda/CqmMeasure.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class CqmMeasure {
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/madie/qrda/CqmMeasure.java:12]: So do you think We can move these models into our package as a part of this story ?

File: src/main/java/gov/cms/madie/services/QrdaService.java (hunk 2 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/services/QrdaService.java]: I thought since it was handling all of this part on server side it wouldn't matter for conversion sake especially since it ends up not using the zone for the final string anyway

File: src/main/java/gov/cms/madie/qrda/CqmMeasure.java (hunk 3 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/qrda/CqmMeasure.java:12]: If it's okay, I'd rather move them as part of another story. I've already been on this story for well over what was projected. It'd also give us a better opportunity to check them over for our purposes and make sure we package them appropriately.

File: src/main/java/gov/cms/madie/services/QrdaClient.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    headers.set(HttpHeaders.CONTENT_TYPE, "application/json");
+    HttpEntity<String> entity = null;
+    try {
+      entity = new HttpEntity<>(mapper.writeValueAsString(dto), headers);
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+    ParameterizedTypeReference<String> responseType = new ParameterizedTypeReference<>() {};
+    try {
+      log.info("fetching the qrda for measure");
```
Comment (nmorasb) [inline src/main/java/gov/cms/madie/services/QrdaClient.java]: can we add the measure ID to this log statement so we have a little more context if troubleshooting

File: src/main/java/gov/cms/madie/services/QrdaClient.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      String result =
+          qrdaRestTemplate.exchange(uri, HttpMethod.PUT, entity, responseType).getBody();
+
+      // Splits the individual documents apart for packaging
+      return Arrays.stream(result.split("</ClinicalDocument>"))
+          .map(s -> s + "\n</ClinicalDocument>")
+          .toList();
+    } catch (Exception ex) {
+      String msg = "An issue occurred while fetching the qrda for measure";
```
Comment (nmorasb) [inline src/main/java/gov/cms/madie/services/QrdaClient.java]: can we add the measure ID to this log statement so we have a little more context if troubleshooting

File: src/main/java/gov/cms/madie/services/TranslationServiceClient.java [hunk 6]
```diff
@@ -68,4 +70,25 @@ public CqlLookups getCqlLookups(QdmMeasure measure, String accessToken) {
...
+    URI uri =
+        URI.create(
+            translatorClientConfig.getBaseUrl() + translatorClientConfig.getRelevantDataElements());
+    HttpHeaders headers = new HttpHeaders();
+    headers.set(HttpHeaders.AUTHORIZATION, accessToken);
+    HttpEntity<Measure> entity = new HttpEntity<>(measure, headers);
+    ParameterizedTypeReference<List<SourceDataCriteria>> responseType =
+        new ParameterizedTypeReference<>() {};
+    try {
+      log.info("fetching the relevant data elements for measure");
```
Comment (nmorasb) [inline src/main/java/gov/cms/madie/services/TranslationServiceClient.java]: can we add the measure ID to this log statement so we have a little more context if troubleshooting

File: src/main/java/gov/cms/madie/services/TranslationServiceClient.java [hunk 7]
```diff
@@ -68,4 +70,25 @@ public CqlLookups getCqlLookups(QdmMeasure measure, String accessToken) {
...
+      return elmTranslatorRestTemplate
+          .exchange(uri, HttpMethod.PUT, entity, responseType)
+          .getBody();
+    } catch (Exception ex) {
+      String msg = "An issue occurred while fetching the relevant data elements for measure";
```
Comment (nmorasb) [inline src/main/java/gov/cms/madie/services/TranslationServiceClient.java]: can we add the measure ID to this log statement so we have a little more context if troubleshooting

File: src/main/java/gov/cms/madie/services/PackagingService.java [hunk 8]
```diff
@@ -67,7 +68,24 @@ public byte[] createMeasurePackage(Measure measure, String accessToken) {
...
-    return "test qrda".getBytes();
+
+    List<String> qrdas = qrdaService.generateQrda((QdmMeasure) measure, accessToken);
+    if (CollectionUtils.isEmpty(qrdas)) {
+      return new byte[0];
+    }
+
+    String zipName = measure.getEcqmTitle() + "-" + measure.getVersion() + "QDM-TestCases";
+
+    log.info("Adding measure package artifacts to the measure package");
```
Comment (nmorasb) [inline src/main/java/gov/cms/madie/services/PackagingService.java]: can we add the measure ID to this log statement so we have a little more context if troubleshooting

File: src/main/java/gov/cms/madie/services/QrdaClient.java (hunk 4 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/services/QrdaClient.java]: fixed

File: src/main/java/gov/cms/madie/services/QrdaClient.java (hunk 5 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/services/QrdaClient.java]: fixed

File: src/main/java/gov/cms/madie/services/TranslationServiceClient.java (hunk 6 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/services/TranslationServiceClient.java]: fixed

File: src/main/java/gov/cms/madie/services/TranslationServiceClient.java (hunk 7 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/services/TranslationServiceClient.java]: fixed

File: src/main/java/gov/cms/madie/services/PackagingService.java (hunk 8 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/services/PackagingService.java]: fixed

File: src/main/java/gov/cms/madie/services/QrdaService.java (hunk 2 above)
Comment (nmorasb) [reply src/main/java/gov/cms/madie/services/QrdaService.java]: I think this might cause issues because, I believe, we're trying to force measurement period to UTC in the browser, and then this will convert that to local/default timezone. Which, if we are not explicitly setting to UTC on the server, would probably be America/New_York since we run in AWS US-East1 region.
This means what we might see is for measurement period start, in the browser they set `01/01/2024`, which I'm seeing in the JSON payload as `2024-01-01T00:00:00.000+00:00`.
Then on the backend, this would be turned into a ZonedDateTime `2023-12-31T19:00-05:00[America/New_York]` and LocalDateTime of `2023-12-31T19:00`.

Comment (nmorasb) [reply src/main/java/gov/cms/madie/services/QrdaService.java]: And this might apply to measurementPeriodEnd logic below, as well as the CqmMeasureMapper logic

File: src/main/java/gov/cms/madie/services/CqmMeasureMapper.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  default MeasurePeriod getMeasurePeriod(QdmMeasure measure) {
+    JSONObject low = new JSONObject();
+    low.put(
+        "value",
+        measure
+            .getMeasurementPeriodStart()
+            .toInstant()
+            .atZone(ZoneId.systemDefault())
```
Comment (nmorasb) [inline src/main/java/gov/cms/madie/services/CqmMeasureMapper.java]: need to confirm how this will react with measurement period coming as UTC from UI

File: src/main/java/gov/cms/madie/services/CqmMeasureMapper.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+            .toLocalDateTime()
+            .format(DateTimeFormatter.ofPattern("yyyyMMddHH")));
+
+    JSONObject high = new JSONObject();
+    high.put(
+        "value",
+        measure
+            .getMeasurementPeriodEnd()
+            .toInstant()
+            .atZone(ZoneId.systemDefault())
```
Comment (nmorasb) [inline src/main/java/gov/cms/madie/services/CqmMeasureMapper.java]: need to confirm how this will react with measurement period coming as UTC from UI

**Review by nmorasb — DISMISSED** [review]

**Review by nmorasb — APPROVED** [review]

File: src/main/java/gov/cms/madie/services/QrdaService.java (hunk 2 above)
Comment (nmorasb) [reply src/main/java/gov/cms/madie/services/QrdaService.java]: @chubert-sb - resolving this for purposes of merge, but let's follow up on this with some dedicated testing and make sure it outputs properly in the export

File: src/main/java/gov/cms/madie/services/CqmMeasureMapper.java (hunk 9 above)
Comment (nmorasb) [reply src/main/java/gov/cms/madie/services/CqmMeasureMapper.java]: @chubert-sb - resolving this for purposes of merge, but let's follow up on this with some dedicated testing and make sure it outputs properly in the export

File: src/main/java/gov/cms/madie/services/CqmMeasureMapper.java (hunk 10 above)
Comment (nmorasb) [reply src/main/java/gov/cms/madie/services/CqmMeasureMapper.java]: @chubert-sb - resolving this for purposes of merge, but let's follow up on this with some dedicated testing and make sure it outputs properly in the export
